File: 1089-duplicate-zeros/1089-duplicate-zeros.java
class Solution {
    
    
    Map<Integer, Integer> findElementAtTail(int[] arr)
    {
        int count = 0;
        int n = arr.length;
        Map<Integer, Integer> mp = new HashMap<>();
        for(int i = 0; i<=n - 1; i++)
        {
            
            mp.put(i, i + count);
            if(arr[i] == 0)
                count++;
        }
        return mp;
    }
    
    
    
    public void duplicateZeros(int[] arr) {
        Map<Integer, Integer> mp = findElementAtTail(arr);
        int n = arr.length;
        int j = n-1;
        
        
        for(int i = n-1; i>=0; i--)
        {
            if(mp.get(i) <= n-1)
            {
                arr[mp.get(i)] = arr[i];
                if(arr[i] == 0 && mp.get(i) + 1 <= n-1)
                    arr[mp.get(i) + 1] = 0;
            }
        }
    }
}

File: 1089-duplicate-zeros/1089-duplicate-zeros.cpp
class Solution {
public:
    
    unordered_map<int, int> findElementAtTail(vector<int> &arr)
    {
        int count = 0;
        int n = arr.size();
        unordered_map<int, int> mp;
        for(int i = 0; i<=n - 1; i++)
        {
            
            mp[i] = i + count;
            if(arr[i] == 0)
                count++;
        }
        return mp;
    }
    
    void duplicateZeros(vector<int>& arr) {
        unordered_map<int, int> mp = findElementAtTail(arr);
        int n = arr.size();
        int j = n-1;
        for(int i = n-1; i>=0; i--)
        {
            if(mp[i] <= n-1)
            {
                arr[mp[i]] = arr[i];
                if(arr[i] == 0 && mp[i] + 1 <= n-1)
                    arr[mp[i] + 1] = 0;
            }
        }
    }
};